File: src/main/java/com/projectswg/holocore/resources/support/data/server_info/loader/SkillLoader.java
/***********************************************************************************
 * Copyright (c) 2019 /// Project SWG /// www.projectswg.com                       *
 *                                                                                 *
 * ProjectSWG is the first NGE emulator for Star Wars Galaxies founded on          *
 * July 7th, 2011 after SOE announced the official shutdown of Star Wars Galaxies. *
 * Our goal is to create an emulator which will provide a server for players to    *
 * continue playing a game similar to the one they used to play. We are basing     *
 * it on the final publish of the game prior to end-game events.                   *
 *                                                                                 *
 * This file is part of Holocore.                                                  *
 *                                                                                 *
 * --------------------------------------------------------------------------------*
 *                                                                                 *
 * Holocore is free software: you can redistribute it and/or modify                *
 * it under the terms of the GNU Affero General Public License as                  *
 * published by the Free Software Foundation, either version 3 of the              *
 * License, or (at your option) any later version.                                 *
 *                                                                                 *
 * Holocore is distributed in the hope that it will be useful,                     *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of                  *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                   *
 * GNU Affero General Public License for more details.                             *
 *                                                                                 *
 * You should have received a copy of the GNU Affero General Public License        *
 * along with Holocore.  If not, see <http://www.gnu.org/licenses/>.               *
 ***********************************************************************************/

package com.projectswg.holocore.resources.support.data.server_info.loader;

import com.projectswg.holocore.resources.support.data.server_info.SdbLoader;
import com.projectswg.holocore.resources.support.data.server_info.SdbLoader.SdbResultSet;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.IOException;
import java.util.*;

public final class SkillLoader extends DataLoader {
	
	private final Map<String, SkillInfo> skillsByName;
	
	SkillLoader() {
		this.skillsByName = new HashMap<>();
	}
	
	@Nullable
	public SkillInfo getSkillByName(String name) {
		return skillsByName.get(name);
	}
	
	@NotNull
	public Collection<SkillInfo> getSkills() {
		return Collections.unmodifiableCollection(skillsByName.values());
	}
	
	@Override
	public final void load() throws IOException {
		try (SdbResultSet set = SdbLoader.load(new File("serverdata/skill/skills.sdb"))) {
			while (set.next()) {
				SkillInfo skills = new SkillInfo(set);
				skillsByName.put(skills.getName(), skills);
			}
		}
	}
	
	@SuppressWarnings("ClassWithTooManyFields")
	public static class SkillInfo {
		
		private final String name;
		private final String parent;
		private final String graphType;
		private final boolean godOnly;
		private final boolean title;
		private final boolean profession;
		private final boolean hidden;
		private final int moneyRequired;
		private final int pointsRequired;
		private final int skillsRequiredCount;
		private final String [] skillsRequired;
		private final String [] preclusionSkills;
		private final String xpType;
		private final int xpCost;
		private final int xpCap;
		private final String [] missionsRequired;
		private final int apprenticeshipsRequired;
		private final String [] statsRequired;
		private final String [] speciesRequired;
		private final String jediStateRequired;
		private final String skillAbility;
		private final String [] commands;
		private final Map<String, Integer> skillMods;
		private final String [] schematicsGranted;
		private final String [] schematicsRevoked;
		private final boolean searchable;
		private final int ender;
		
		public SkillInfo(SdbResultSet set) {
			this.name = set.getText("name");
			this.parent = set.getText("parent");
			this.graphType = set.getText("graph_type");
			this.godOnly = set.getBoolean("god_only");
			this.title = set.getBoolean("is_title");
			this.profession = set.getBoolean("is_profession");
			this.hidden = set.getBoolean("is_hidden");
			this.moneyRequired = (int) set.getInt("money_required");
			this.pointsRequired = (int) set.getInt("points_required");
			this.skillsRequiredCount = (int) set.getInt("skills_required_count");
			this.skillsRequired = splitCsv(set.getText("skills_required"));
			this.preclusionSkills = splitCsv(set.getText("preclusion_skills"));
			this.xpType = set.getText("xp_type");
			this.xpCost = (int) set.getInt("xp_cost");
			this.xpCap = (int) set.getInt("xp_cap");
			this.missionsRequired = splitCsv(set.getText("missions_required"));
			this.apprenticeshipsRequired = (int) set.getInt("apprenticeships_required");
			this.statsRequired = splitCsv(set.getText("stats_required"));
			this.speciesRequired = splitCsv(set.getText("species_required"));
			this.jediStateRequired = set.getText("jedi_state_required");
			this.skillAbility = set.getText("skill_ability");
			this.commands = splitCsv(set.getText("commands"));
			this.skillMods = createSkillModMap(splitCsv(set.getText("skill_mods")));
			this.schematicsGranted = splitCsv(set.getText("schematics_granted"));
			this.schematicsRevoked = splitCsv(set.getText("schematics_revoked"));
			this.searchable = set.getBoolean("searchable");
			this.ender = (int) set.getInt("ender");
		}
		
		public String getName() {
			return name;
		}
		
		public String getParent() {
			return parent;
		}
		
		public String getGraphType() {
			return graphType;
		}
		
		public boolean isGodOnly() {
			return godOnly;
		}
		
		public boolean isTitle() {
			return title;
		}
		
		public boolean isProfession() {
			return profession;
		}
		
		public boolean isHidden() {
			return hidden;
		}
		
		public int getMoneyRequired() {
			return moneyRequired;
		}
		
		public int getPointsRequired() {
			return pointsRequired;
		}
		
		public int getSkillsRequiredCount() {
			return skillsRequiredCount;
		}
		
		public String[] getSkillsRequired() {
			return skillsRequired.clone();
		}
		
		public String[] getPreclusionSkills() {
			return preclusionSkills.clone();
		}
		
		public String getXpType() {
			return xpType;
		}
		
		public int getXpCost() {
			return xpCost;
		}
		
		public int getXpCap() {
			return xpCap;
		}
		
		public String[] getMissionsRequired() {
			return missionsRequired.clone();
		}
		
		public int getApprenticeshipsRequired() {
			return apprenticeshipsRequired;
		}
		
		public String[] getStatsRequired() {
			return statsRequired.clone();
		}
		
		public String[] getSpeciesRequired() {
			return speciesRequired.clone();
		}
		
		public String getJediStateRequired() {
			return jediStateRequired;
		}
		
		public String getSkillAbility() {
			return skillAbility;
		}
		
		public String[] getCommands() {
			return commands.clone();
		}
		
		public Map<String, Integer> getSkillMods() {
			return Collections.unmodifiableMap(skillMods);
		}
		
		public String[] getSchematicsGranted() {
			return schematicsGranted.clone();
		}
		
		public String[] getSchematicsRevoked() {
			return schematicsRevoked.clone();
		}
		
		public boolean isSearchable() {
			return searchable;
		}
		
		public int getEnder() {
			return ender;
		}
		
		private static String [] splitCsv(String str) {
			if (str.isEmpty())
				return new String[0];
			else if (str.indexOf(',') == -1)
				return new String[]{str};
			return str.split(",");
		}
		
		private static Map<String, Integer> createSkillModMap(String [] elements) {
			Map<String, Integer> skillMods = new HashMap<>();
			for (String element : elements) {
				String [] split = element.split("=", 2);
				if (split.length < 2)
					continue;
				skillMods.put(split[0], Integer.valueOf(split[1]));
			}
			return skillMods;
		}
		
	}
}

File: src/main/java/com/projectswg/holocore/resources/support/data/server_info/loader/SkillLoader.kt
/***********************************************************************************
 * Copyright (c) 2024 /// Project SWG /// www.projectswg.com                       *
 *                                                                                 *
 * ProjectSWG is the first NGE emulator for Star Wars Galaxies founded on          *
 * July 7th, 2011 after SOE announced the official shutdown of Star Wars Galaxies. *
 * Our goal is to create an emulator which will provide a server for players to    *
 * continue playing a game similar to the one they used to play. We are basing     *
 * it on the final publish of the game prior to end-game events.                   *
 *                                                                                 *
 * This file is part of Holocore.                                                  *
 *                                                                                 *
 * --------------------------------------------------------------------------------*
 *                                                                                 *
 * Holocore is free software: you can redistribute it and/or modify                *
 * it under the terms of the GNU Affero General Public License as                  *
 * published by the Free Software Foundation, either version 3 of the              *
 * License, or (at your option) any later version.                                 *
 *                                                                                 *
 * Holocore is distributed in the hope that it will be useful,                     *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of                  *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                   *
 * GNU Affero General Public License for more details.                             *
 *                                                                                 *
 * You should have received a copy of the GNU Affero General Public License        *
 * along with Holocore.  If not, see <http://www.gnu.org/licenses/>.               *
 ***********************************************************************************/
package com.projectswg.holocore.resources.support.data.server_info.loader

import com.projectswg.holocore.resources.support.data.server_info.SdbLoader
import com.projectswg.holocore.resources.support.data.server_info.SdbLoader.SdbResultSet
import java.io.File
import java.io.IOException
import java.util.*

class SkillLoader internal constructor() : DataLoader() {
	private val skillsByName: MutableMap<String, SkillInfo> = HashMap()

	fun getSkillByName(name: String): SkillInfo? {
		return skillsByName[name]
	}

	val skills: Collection<SkillInfo>
		get() = Collections.unmodifiableCollection(skillsByName.values)

	@Throws(IOException::class)
	override fun load() {
		SdbLoader.load(File("serverdata/skill/skills.sdb")).use { set ->
			while (set.next()) {
				val skills = SkillInfo(set)
				skillsByName[skills.name] = skills
			}
		}
	}

	class SkillInfo(set: SdbResultSet) {
		val name: String = set.getText("name")
		val parent: String = set.getText("parent")
		val graphType: String = set.getText("graph_type")
		val isGodOnly: Boolean = set.getBoolean("god_only")
		val isTitle: Boolean = set.getBoolean("is_title")
		val isProfession: Boolean = set.getBoolean("is_profession")
		val isHidden: Boolean = set.getBoolean("is_hidden")
		val moneyRequired: Int = set.getInt("money_required").toInt()
		val pointsRequired: Int = set.getInt("points_required").toInt()
		val skillsRequiredCount: Int = set.getInt("skills_required_count").toInt()
		val skillsRequired: List<String> = splitCsv(set.getText("skills_required"))
		val preclusionSkills: List<String> = splitCsv(set.getText("preclusion_skills"))
		val xpType: String = set.getText("xp_type")
		val xpCost: Int = set.getInt("xp_cost").toInt()
		val xpCap: Int = set.getInt("xp_cap").toInt()
		val missionsRequired: List<String> = splitCsv(set.getText("missions_required"))
		val apprenticeshipsRequired: Int = set.getInt("apprenticeships_required").toInt()
		private val statsRequired: List<String> = splitCsv(set.getText("stats_required"))
		private val speciesRequired: List<String> = splitCsv(set.getText("species_required"))
		val jediStateRequired: String = set.getText("jedi_state_required")
		val skillAbility: String = set.getText("skill_ability")
		val commands: List<String> = splitCsv(set.getText("commands"))
		val skillMods: Map<String, Int> = createSkillModMap(splitCsv(set.getText("skill_mods")))
		val schematicsGranted: List<String> = splitCsv(set.getText("schematics_granted"))
		val schematicsRevoked: List<String> = splitCsv(set.getText("schematics_revoked"))
		val isSearchable: Boolean = set.getBoolean("searchable")
		val ender: Int = set.getInt("ender").toInt()

		companion object {
			private fun splitCsv(str: String): List<String> {
				if (str.isEmpty()) return listOf()
				else if (str.indexOf(',') == -1) return listOf(str)
				return str.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toList()
			}

			private fun createSkillModMap(elements: List<String>): Map<String, Int> {
				val skillMods: MutableMap<String, Int> = HashMap()
				for (element in elements) {
					val split = element.split("=".toRegex(), limit = 2).toTypedArray()
					if (split.size < 2) continue
					skillMods[split[0]] = split[1].toInt()
				}
				return skillMods
			}
		}
	}
}
